R/prop_diff.R

```diff
@@ -287,17 +290,140 @@ prop_diff_cmh <- function(rsp,
+#'   implementation follows closely the one proposed by `Yan` and `Su` (2010).
+#'   Weights can be estimated from the heuristic proposed in
+#'   [prop_strat_wilson()] or from CMH-derived weights (see [prop_diff_cmh()]).
+#'
+#' @param strata (`factor`)\cr
+#'   with one level per stratum and same length as `rsp`.
+#' @param weights_method (`string`) \cr
+#'   it can be one of `c("cmh", "heuristic")` and directs the way weights are estimated.
+#'
+#' @examples
+#' # Stratified Newcombe confidence interval
+#'
+#' set.seed(2)
+#' data_set <- data.frame(
+#'   "rsp" = sample(c(TRUE, FALSE), 100, TRUE),
+#'   "f1" = sample(c("a", "b"), 100, TRUE),
+#'   "f2" = sample(c("x", "y", "z"), 100, TRUE),
+#'   "grp" = sample(c("Placebo", "Treatment"), 100, TRUE),
+#'   stringsAsFactors = TRUE
+#' )
+#'
+#' prop_diff_strat_nc(
+#'   rsp = data_set$rsp, grp = data_set$grp, strata = interaction(data_set[2:3]),
+#'   weights_method = "cmh",
+#'   conf_level = 0.90
+#' )
+#'
+#' prop_diff_strat_nc(
+#'   rsp = data_set$rsp, grp = data_set$grp, strata = interaction(data_set[2:3]),
+#'   weights_method = "wilson_h",
+#'   conf_level = 0.90
+#' )
+#'
+#' @references
+#' \itemize{
+#'   \item `Yan, Xin, and Xiao Gang Su. 2010. “Stratified Wilson and Newcombe Confidence Intervals for Multiple Binomial Proportions.” Statistics in Biopharmaceutical Research 2 (3): 329–35.`
+#' }
+#'
+#' @export
+prop_diff_strat_nc <- function(rsp,
+                               grp,
+                               strata,
+                               weights_method = c("cmh", "wilson_h")[1],
+                               conf_level = 0.95,
+                               correct = FALSE) {
+
+  # Checks
+  checkmate::assert_choice(weights_method, choices = c("cmh", "wilson_h"))
+  grp <- as_factor_keep_attributes(grp)
+  strata <- as_factor_keep_attributes(strata)
+  check_diff_prop_ci(
+    rsp = rsp, grp = grp, conf_level = conf_level, strata = strata
+  )
+  checkmate::assert_number(conf_level, lower = 0, upper = 1)
+  checkmate::assert_flag(correct)
+  if (any(tapply(rsp, strata, length) < 5)) {
+    warning("Less than 5 observations in some strata.")
+  }
+
+  # Splitting
+  rsp_by_grp <- split(rsp, f = grp)
+  strata_by_grp <- split(strata, f = grp)
+
+  # Finding the weights
+  if (identical(weights_method, "cmh")) {
+    weights <- prop_diff_cmh(rsp = rsp, grp = grp, strata = strata)$weights
+  } else if (identical(weights_method, "wilson_h")) {
+    weights <- prop_strat_wilson(rsp, strata, conf_level = conf_level, correct = correct)$weights
+  }
+
+  # Calculating L and U per group
+  strat_wilson_by_grp <- Map(
+    prop_strat_wilson,
+    rsp = rsp_by_grp,
+    strata = strata_by_grp,
+    weights = list(weights, weights),
+    conf_level = conf_level,
+    correct = correct
+  )
+
+  ci_ref <- strat_wilson_by_grp[[1]]
+  ci_trt <- strat_wilson_by_grp[[2]]
+  l1 <- array(ci_ref$conf.int[1]) # array() loses the names
```

Comment:
better 
```suggestion
  l_ref <- as.numeric(ci_ref$conf.int[1])
```
etc. below - better name and `as.numeric()` does the same but is not confusing that there might be multiple dimensions

Reply:
that is way better, should I also add `l_trt` for the other values (i.e. `l2` and `*2`)?

Reply:
I did it for all in the function. I did not extend this to `cmh`

Reply:
perfect thx